system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1331,354 @@ handle_run_triggers (FlatpakSystemHelper   *object,
   return TRUE;
 }
 
+static gboolean
+get_connection_uid (GDBusMethodInvocation *invocation, uid_t *uid_out, GError **error)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  const gchar *sender = g_dbus_method_invocation_get_sender (invocation);
+  g_autoptr(GVariant) dict = NULL;
+  g_autoptr(GVariant) unixUserId = NULL;
+  g_autoptr(GVariant) credentials = NULL;
+  uid_t uid;
+
+   credentials = g_dbus_connection_call_sync (connection,
+                                             "org.freedesktop.DBus",
+                                             "/org/freedesktop/DBus",
+                                             "org.freedesktop.DBus",
+                                             "GetConnectionCredentials",
+                                             g_variant_new ("(s)", sender),
+                                             G_VARIANT_TYPE ("(a{sv})"), G_DBUS_CALL_FLAGS_NONE, G_MAXINT, NULL, error);
+  if (credentials == NULL)
+    return FALSE;
+
+   dict = g_variant_get_child_value (credentials, 0);
+  unixUserId = g_variant_lookup_value (dict, "UnixUserID", G_VARIANT_TYPE_UINT32);
+
+   if (unixUserId == NULL)
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                   "Failed to query UnixUserId for the bus name: %s", sender);
+      return FALSE;
+    }
+
+   *uid_out = g_variant_get_uint32 (unixUserId);
+  return TRUE;
+}
+
+
+static void
+revokefs_child_setup (gpointer user_data)
+{
+  struct passwd *passwd = getpwnam("flatpak");
+  setgid (passwd->pw_gid);
+  setuid (passwd->pw_uid);
+
+  close (GPOINTER_TO_INT(user_data));
+}
+
+static void
+name_vanished_cb (GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+  OngoingPull *pull = (OngoingPull *) user_data;
+
+  ongoing_pull_free (pull);
+}
+
+static OngoingPull *
+ongoing_pull_new (FlatpakSystemHelper   *object,
+                  GDBusMethodInvocation *invocation,
+                  const gchar           *ref,
+                  const gchar           *commit,
+                  uid_t                  uid,
+                  gchar                 *revokefs_fuse_path,
+                  gchar                 *src,
```

Comment:
If ownership of these parameters is not transferred, they should be `const gchar *`